src/r3d.imageio/r3dinput.cpp

```diff
@@ -0,0 +1,441 @@
+
+    m_spec = ImageSpec(width, height, m_channels, TypeDesc::UINT16);
+    m_spec.attribute("FramesPerSecond", TypeFloat, &m_fps);
+    m_spec.attribute("oiio:Movie", true);
+    m_spec.attribute("oiio:subimages", int(m_frames));
+    m_spec.attribute("oiio:BitsPerSample", 16);
+
+    newspec         = m_spec;
+    m_next_scanline = 0;
+    return true;
+}
+
+
+
+void
+R3dInput::read_frame(int pos)
+{
+    DBG std::cout << "R3dInput::read_frame(" << pos << ")\n";
+
+    if (m_last_decoded_pos + 1 != pos) {
+        seek(pos);
+    }
+
+    R3DSDK::DecodeStatus status = m_clip->DecodeVideoFrame(pos, m_job);
+    if (status != R3DSDK::DSDecodeOK) {
+        DBG std::cout << "Failed to decode frame " << pos << "\n";
+    }
+
+    m_last_search_pos  = pos;
+    m_last_decoded_pos = pos;
+    m_read_frame       = true;
+    m_next_scanline    = 0;
+}
+
+
+
+bool
+R3dInput::seek_subimage(int subimage, int miplevel)
+{
+    DBG std::cout << "R3dInput::seek_subimage(" << subimage << ", " << miplevel
+                  << ")\n";
+
+    if (subimage < 0 || subimage >= m_nsubimages || miplevel > 0) {
+        return false;
+    }
+    if (subimage == m_subimage) {
+        return true;
+    }
+    m_subimage   = subimage;
+    m_read_frame = false;
+    return true;
+}
+
+
+
+bool
+R3dInput::read_native_scanline(int subimage, int miplevel, int y, int /*z*/,
+                               void* data)
+{
+    DBG std::cout << "R3dInput::read_native_scanline(, , " << y << ")\n";
+
+    lock_guard lock(*this);
+    if (!seek_subimage(subimage, miplevel))
+        return false;
+    if (!m_read_frame) {
+        read_frame(m_subimage);
+    }
+    if (y < 0 || y >= m_spec.height)  // out of range scanline
+        return false;
+    if (m_next_scanline > y) {
+        // User is trying to read an earlier scanline than the one we're
+        // up to.  Easy fix: close the file and re-open.
+        // Don't forget to save and restore any configuration settings.
+        ImageSpec configsave;
+        if (m_config)
+            configsave = *m_config;
+        ImageSpec dummyspec;
+        int subimage = current_subimage();
+        if (!close() || !open(m_filename, dummyspec, configsave)
+            || !seek_subimage(subimage, 0))
+            return false;  // Somehow, the re-open failed
+        OIIO_DASSERT(m_next_scanline == 0 && current_subimage() == subimage);
+    }
+
+    uint16_t *r, *g, *b;
+    uint16_t* d          = (uint16_t*)data;
+    size_t scanline_size = m_spec.width * sizeof(uint16_t);
+    size_t plane_size    = scanline_size * m_spec.height;
+
+    r = (uint16_t*)((uint8_t*)m_image_buffer + y * scanline_size);
+    g = (uint16_t*)((uint8_t*)m_image_buffer + y * scanline_size + plane_size);
+    b = (uint16_t*)((uint8_t*)m_image_buffer + y * scanline_size
+                    + 2 * plane_size);
+
+    for (int x = 0; x < m_spec.width; x++) {
+        *d++ = r[x];
+        *d++ = g[x];
+        *d++ = b[x];
+    }
```

Comment:
I think you can use OIIO::copy_image for this. It's declared in imageio.h, and in tiffinput.cpp you can find examples of how it's used to copy "planar" to "contiguous" striding.

Reply:
Thanks. Will try and use that.